lib/transfer-memory-allocator.js

```diff
@@ -0,0 +1,544 @@
+    set size(value) {
+        this[INTERNAL].size = value;
+    }
+
+    /**
+     * Get the start index from the buffer pool (lowest index/memory
+     * location usable for this Transfer Memory Block)
+     * @returns {Integer} Start index of the allocated buffer 
+     * (memory block) in the buffer pool
+     */
+    get startIndex() {
+        return this[INTERNAL].bufferPoolStartIndex;
+    }
+
+    /**
+     * the start index from the buffer pool (lowest index/memory
+     * location usable for this Transfer Memory Block). 
+     * Used by Transfer Memory Block allocator.
+     * -1 means `unallocated` (or `deallocated`)
+     * @param {Integer} value Start index in buffer pool
+     */
+    set startIndex(value) {
+        this[INTERNAL].bufferPoolStartIndex = value;
+    }
+
+    /**
+     * Get the end index from the buffer pool (highest index/memory
+     * location usable for this Transfer Memory Block)
+     * @returns {Integer} End index of the allocated buffer 
+     * (memory block) in the buffer pool
+     */
+    get endIndex() {
+        return this.startIndex + this.size - 1;
+    }
+
+    /**
+     * Get the buffer representing the memory block from the pooled
+     * buffer allocated for this Transfer Memory Block.
+     * Modifying this memory block (buffer slice) will modify the memory 
+     * in the original Buffer because the allocated memory of this memory block 
+     * and the original buffer pool overlap.
+     */
+    get buffer() {
+        return this[INTERNAL].referencedMemoryFromBufferPool;
+    }
+
+    /**
+     * Set the buffer to use for memory access. 
+     * Used by Transfer Memory Block allocator
+     * @param {Buffer} value Buffer to use for memory access 
+     */
+    set buffer(value) {
+        return this[INTERNAL].referencedMemoryFromBufferPool = value;
+    }
+}
+
+// ------------------------------------------------------------------------------
+const PRIVATE = Symbol("private");
+/**
+ * Keeps track of memory usable for transfers, and allocates available 
+ * memory as needed/when requested by transfers
+ */
+class TransferMemoryBuffer {
+    /**
+     * Constructs a transfer memory buffer
+     * @param {*} suggestedSize Suggested size for the memory reserved 
+     * for buffering transfer data, in bytes. The memory allocator will
+     * try to allocate memory with size `suggestedSize`, but if not 
+     * enough memory is available, 80% of available memory will be 
+     * allocated instead
+     */
+    constructor(suggestedSize) {
+        this[PRIVATE] = {};
+
+        let bufferSize;
+        if (suggestedSize) {
+            console.log(`Suggested Transfer Memory Buffer pool size is ${suggestedSize} bytes`);
+            bufferSize = parseInt(suggestedSize, 10);
+        } else {
+            console.log(`Using default size for Transfer Memory Buffer: ${suggestedSize} bytes`);
+            bufferSize = DEFAULT_TRANSFER_BUFFER_POOL_SIZE;
+        }
+        if (suggestedSize <= 0) {
```

Comment:
why not just use the default if the suggested size is <0?

Reply:
Because the caller made a mistake regarding the memory they want to use. This is an error, other languages would just complain with a `SEGFAULT`. Since this is an error dealing with memory, I prefer to throw rather than silently use a default and possibly allocate memory that should not be.